### DISCUSSION TITLE
customizing generated model/service

### DISCUSSION THREAD

**Discussion by d-a-s** [odata2ts/odata2ts#318] [Ideas]
I've run into a situation where I need to inject additional functionality into `EntityTypeServiceV4` and `EntitySetServiceV4`. It's quite easy to do this by importing and extending these classes in my local code, and then manually editing the imports at the top of the generated service file. Something like:
```
- import { ODataService, EntityTypeServiceV4, EntitySetServiceV4, CollectionServiceV4 } from "@odata2ts/odata-service";
+ import { ODataService, CollectionServiceV4 } from "@odata2ts/odata-service";
+ import { EntityTypeServiceV4, EntitySetServiceV4 } from "./odata-service-overrides";
```
(As a side note the functionality I'm adding is a version of `.query()` which does some additional functionality relating to server-side data table filtering)

I don't currently see any option/hook to do this sort of override. Looking at `addServiceObject` (called by `generateEntityTypeService` and `generateEntityCollectionService`, it only seems to allow for adding the service object via `this.libs.service`, which in `getImportDeclarations` always uses the module name from `LIB_MODULES`.

We have already found a short-term work-around to make the necessary change. But I wanted to know if you'd be open to adding a way to do this during generation? If so, I'm happy to do some or all of the coding.

**Comment by texttechne** [comment]
Hi @d-a-s,

I understand that you want to add custom functionality to the generated services.
However, you don't really explain your problem. This is actually the only hint:

> As a side note the functionality I'm adding is a version of .query() which does some additional functionality relating to server-side data table filtering

So could you please explain your issue the other way around: Instead of focusing on how you solved it, expand on what your exact problem is and what you are trying to achieve. That's not a "side note", but the real meat of the issue 😉 
And code examples would be really helpful!

**Reply by texttechne** [reply]
Hi @d-a-s,
I don't want to be disencouraging and I really appreciate any help. So let me give you some background regarding my request for more information.

On the one hand, the way you solved your problem isn't sound from a library perspective: If odata2ts would allow you to switch out those imports, you could do literally anything without obeying to any sort of contract / interface (not even that you should extend `EntityTypeService` for example). This would in turn mean that you would carefully need to check your code at runtime and, furthermore, any library change down the road could break your code (runtime checks would be needed again). 

On the other hand, and I think most importantly, I didn't envisage any extension of the generated services since they simply reflect OData functionality. From this standpoint either something is missing in the generated artefacts or your problem could be solved in a different way. And, of course, I could be wrong with this hypothesis. However, I cannot judge this without more information.

**Comment by d-a-s** [comment]
Thanks for following up, sorry for taking so long to reply; I got busy with some project deadlines and then forgot about it!

The end goal is a tighter integration with tanstack/query, and also with a server-side dynamic filtering system for a "data-table" integration. To make these work in an efficient manner, it was quite useful to integrate with some non-exposed portions of the `EntityTypeService` and `EntitySetService`

Here are the obstacles we ran into:
- Access to the generated url, to use as a cache key. (We use tanstack/query)
- Adding dynamic `queryBuilder` overrides. In our data tables, we have the column name in our table definition, so it's relatively easy to do something like `(b,q) => b.filter(q[key].contains(value))` (greatly simplified). It mostly comes down to being able to apply multiple `queryFn`s. to the `ODataQueryBuilder` before the `builder.build()` is executed.
- Full type support for return array: originally I was doing all of this in an external function rather than extending the classes, but I couldn't get the return type to work. (I was also having to access the private `this.__base` to do it that way, which isn't really ideal.)

Ultimately the only way found to accomplish all of these goals was to extend the class. Here is a greatly simplified example:
```
class EntitySetServiceV4<in out ClientType extends ODataHttpClient, T, EditableT, Q extends QueryObjectModel, EIdType>
  extends EntitySetServiceV4Base<ClientType, T, EditableT, Q, EIdType> {

  useQuery<ReturnType extends Partial<T> = T>(args: InputArgsSet<ClientType, Q>): QryTypePag<ReturnType> {
    const { client, qResponseType, applyQueryBuilder, getDefaultHeaders } = this.__base;
    const [dtFilter, dtSort, dtPage] = buildDatatableFilters();
    const url = applyQueryBuilder(combineQueries(
      dtSort, queryFn, dtFilter, dtPage, b=> b.count(true)
    ));
    return useQuery({
      queryKey: [url],
      queryFn: client.get(url, requestConfig, getDefaultHeaders()),
    });
  }
}
```
We're currently using vite aliases and tsconfig paths to replace `@odata2ts/odata-service` with our local extended version.

**Reply by texttechne** [reply]
I think we already talked about this, but I cannot find any corresponding issue or discussion...

Anyways, I would like to tackle this step by step without providing an option to extend the base service classes.

### URL as Cache Key
I intend to refactor the service operations (like `query`) to have an intermediate step.
Along these lines:
```ts
const request = mainService.people().query();
const response = request.execute();
```
This would then allow for more flexibility regarding the response, e.g. getting the data directly (`request.getData()` instead of currently `query().data`) or applying response converters, etc. This would also allow you to retrieve the URL from this request object.

**Reply by texttechne** [reply]
And now I would have more questions regarding the problems you've listed.

### Problem 2
>Adding dynamic queryBuilder overrides. In our data tables, we have the column name in our table definition, so it's relatively easy to do something like (b,q) => b.filter(q[key].contains(value)) (greatly simplified). It mostly comes down to being able to apply multiple queryFns. to the ODataQueryBuilder before the builder.build() is executed.

Sorry, but I don't quite get it. 
Is `q[key]` a problem? 

I also dimly remember that you wanted to store the builder and then just change filtering or sorting. Is that the problem here?
I wouldn't recommend that but instead use a factory function to create the builder every time you need it and then apply the filtering and sorting for the given request. Should be easier as you always perform the same action: apply your filters and your sorting...

### Problem 3
> Full type support for return array: originally I was doing all of this in an external function rather than extending the classes, but I couldn't get the return type to work. (I was also having to access the private this.__base to do it that way, which isn't really ideal.)

Maybe not a problem if you wouldn't use an external function? I mean, odata2ts gives you the right typing...

**Comment by d-a-s** [comment]
> I intend to refactor the service operations (like query) to have an intermediate step.

That would be very helpful; that change alone would greatly simplify what I'm doing.

> Problem 2: Adding dynamic queryBuilder overrides. ... Is `q[key]` a problem?

No, it's not a problem. As I said that's a greatly simplified example. Obviously in my production code I'm checking for things like whether `key` exists in the `qType`, and whether it has a `contains` method, etc.

Also, I'm not asking for any code changes related to that. With your above proposal, I can handle this part of it in my own code.

And yes, I am basically doing what you suggest, with a "factory function".

> Problem 3: Maybe not a problem if you wouldn't use an external function

I have so many pages where I'm doing server-side filtering data tables, I really need to modularize that functionality. But with your proposed refactor, I think it's much more likely that implicit typing will work.

**Comment by texttechne** [comment]
The command pattern has been implemented in recent versions. Hope it helps.
